src/NServiceBus.Core/Recoverability/Faults/MoveFaultsToErrorQueueBehavior.cs

```diff
@@ -8,60 +9,95 @@ namespace NServiceBus
 
     class MoveFaultsToErrorQueueBehavior : ForkConnector<ITransportReceiveContext, IFaultContext>
     {
-        public MoveFaultsToErrorQueueBehavior(CriticalError criticalError, string errorQueueAddress, string localAddress)
+        public MoveFaultsToErrorQueueBehavior(CriticalError criticalError, string errorQueueAddress, string localAddress, TransportTransactionMode transportTransactionMode, FailureInfoStorage failureInfoStorage)
         {
             this.criticalError = criticalError;
             this.errorQueueAddress = errorQueueAddress;
             this.localAddress = localAddress;
+            this.transportTransactionMode = transportTransactionMode;
+            this.failureInfoStorage = failureInfoStorage;
         }
 
+        bool RunningWithTransactions => transportTransactionMode != TransportTransactionMode.None;
+
         public override async Task Invoke(ITransportReceiveContext context, Func<Task> next, Func<IFaultContext, Task> fork)
         {
+            var message = context.Message;
+
+            var failureInfo = failureInfoStorage.GetFailureInfoForMessage(message.MessageId);
+
+            if (failureInfo.MoveToErrorQueue)
+            {
+                await MoveMessageToErrorQueue(context, fork, message, failureInfo.Exception).ConfigureAwait(false);
+
+                return;
+            }
+
             try
             {
                 await next().ConfigureAwait(false);
             }
-            catch (Exception exception)
+            catch (Exception ex)
             {
-                try
+                if (RunningWithTransactions)
                 {
-                    var message = context.Message;
+                    failureInfoStorage.MarkForMovingToErrorQueue(message.MessageId, ExceptionDispatchInfo.Capture(ex));
 
-                    Logger.Error($"Moving message '{message.MessageId}' to the error queue because processing failed due to an exception:", exception);
+                    context.AbortReceiveOperation();
+                }
+                else
```

Comment:
Persoanlly I'd love to see this as a dedicated behavior. This would make the behavior simpler. But this is not a blocker. But any ideas whether there is anything which would stop us from doing it?